libyul/optimiser/Suite.cpp

```diff
@@ -429,6 +246,60 @@ map<char, string> const& OptimiserSuite::stepAbbreviationToNameMap()
 	return lookupTable;
 }
 
+void OptimiserSuite::runSequence(string const& _stepAbbreviations, Block& _ast)
+{
+	string input = _stepAbbreviations;
+	boost::remove_erase(input, ' ');
+	boost::remove_erase(input, '\n');
+
+	bool insideLoop = false;
+	for (char abbreviation: input)
+		switch (abbreviation)
+		{
+		case '(':
+			assertThrow(!insideLoop, OptimizerException, "Nested parentheses not supported");
+			insideLoop = true;
+			break;
+		case ')':
+			assertThrow(insideLoop, OptimizerException, "Unbalanced parenthesis");
+			insideLoop = false;
+			break;
+		default:
+			assertThrow(
+				stepAbbreviationToNameMap().find(abbreviation) != stepAbbreviationToNameMap().end(),
+				OptimizerException,
+				"Invalid optimisation step abbreviation"
+			);
+		}
+	assertThrow(!insideLoop, OptimizerException, "Unbalanced parenthesis");
+
+	auto abbreviationsToSteps = [](string const& _fullSequence, size_t _begin, size_t _end) -> vector<string>
+	{
+		vector<string> steps;
+		for (char abbreviation: _fullSequence.substr(_begin, _end - _begin))
+			steps.push_back(stepAbbreviationToNameMap().at(abbreviation));
+
+		return steps;
+	};
+
+	// The sequence has now been validated and must consist of pairs of segments that look like this: `aaa(bbb)`
```

Comment:
Oh wow, I thought there is only at most one pair of parentheses.

Reply:
I guess this means that the state machine version must indeed have been less readable :)